# Questões  <a name="questoes"></a>
1. [Alturas](#alturas)
2. [IMC](#imc)
3. [Endividamento](#endividamento)
4. [Valores especiais](#4)

In [1]:
"""
Imports
"""
import numpy as np

# 1. Alturas <a name="alturas"></a>

<div style="text-align: right"
     
[Voltar ao índice](#questoes)

A) No trecho de código abaixo, crie um ndarray chamado ```altura_em_centimetros``` transformando a lista ```lista_de_alturas``` em um ndarray do numpy com a função ```np.array()```  
B) Crie um outro objeto ```altura_em_metros```, com os valores de ```altura_em_centimetros``` convertidos para metros.

In [2]:
"""
A) Tranformando lista em ndarray
""" 
lista_de_centimetros = list(range(170, 190, 5))
altura_em_centimetros = np.array(lista_de_centimetros)
altura_em_centimetros

array([170, 175, 180, 185])

In [3]:
"""
B) Convertendo altura em metros
"""
altura_em_metros = altura_em_centimetros/100
altura_em_metros

array([1.7 , 1.75, 1.8 , 1.85])

# 2. IMC <a name="imc"></a>

<div style="text-align: right"
     
[Voltar ao índice](#questoes)

A) Considere que pesos em Kg dessas pessoas, na mesma ortem, estão na lista pesos ```lista_pesos = [70, 75, 80, 85]```. Crie um **ndarray** chamado ```pesos``` com a função ```np.array()``` que contenha esses valores.  
B) Utilizando o objeto que contém as alturas em metros e esse objeto que contém os respectivos pesos em quilos, calcule o IMC desses indivíduos utilizando a aritmética de arrays e guarde os resultados em um objeto chamado ```imc```.

In [4]:
"""
A) Tranformando lista em ndarray
"""
lista_pesos = [70, 75, 80, 85]
pesos = np.array(lista_pesos)
pesos

array([70, 75, 80, 85])

In [ ]:
"""
B) Calculando o IMC dessas pessoas
"""
imc = pesos / (altura_em_metros ** 2)
imc

# 3. Endividamento <a name="endividamento"></a>

<div style="text-align: right"
     
[Voltar ao índice](#questoes)

Cálculos de novas variáveis como endividamento total e comprometimento de renda são essenciais para a construção de modelos financeiros em ciência de dados. Áreas não financeiras terão cálculos semelhantes também. Vamos praticar:

Considere que o seguinte ndarray contém os dados de 4 pessoas, total a ser pago a empréstimos mensalmente e renda familiar:

| custo fixo | dívida financeira | renda familiar |
|:----:|:----:|:---|
| 3000  | 1000 | 6000 |
| 2500  | 2500 | 5500 |
| 1000  | 3000 | 7000 |
| 10000 | 5000 | 16000 |

i. Transforme a lista de listas ```dados_financeiros``` no ndarray ```nd_financeiros```.
> ``` dados_financeiros[[3000, 2500, 1000, 10000],[1000, 2500, 3000, 5000],[6000, 5500, 7000, 16000]] ```

ii. Utilize o método ```.transpose ``` e certifique-se de que esse ndarray tenha uma linha por indivíduo e uma coluna por informação. Utilizando a indexação do numpy, imprima no output a segunda linha do array, depois a segunda coluna.

iii. Pratique aritmética com nearrays e calcule o endividamento total como:
$$endividamento\hspace{.2cm}total = \frac{custo \hspace{.2cm}fixo + dívida\hspace{.2cm}financeira}{renda\hspace{.2cm}familiar}$$
Guarde os resultados em uma variável chamada ```endividamento_total``` e verifique os resultados imprimindo o conteúdo dessa variável no output.

iv. Considere que há um erro de digitação que precisa ser corrigido: 3o indivíduo na verdade não possui renda familiar de R\\$7.000,00, mas sim R\\$ 10.000,00. Corrija esse valor e refaça os cálculos.

In [6]:
"""
A) e B) Tranformando lista em ndarray ja utilizando o metodo .transpose para que seja uma linha por indivíduo e uma coluna por informação
"""
dados_financeiros = [[3000, 2500, 1000, 10000],[1000, 2500, 3000, 5000],[6000, 5500, 7000, 16000]]
nd_financeiros = np.transpose(dados_financeiros)
nd_financeiros

array([[ 3000,  1000,  6000],
       [ 2500,  2500,  5500],
       [ 1000,  3000,  7000],
       [10000,  5000, 16000]])

In [7]:
"""
C) Calculando o endividamento total 
"""
endividamento_total=(nd_financeiros[:,0]+nd_financeiros[:,1])/nd_financeiros[:,2]
endividamento_total*100

array([66.66666667, 90.90909091, 57.14285714, 93.75      ])

In [8]:
"""
D) Corrigindo um valor específico
"""
print('Antes:\n{}'.format(nd_financeiros))
nd_financeiros_corrigido = nd_financeiros
nd_financeiros_corrigido[2,2] = 10000
print('Depois:\n{}'.format(nd_financeiros_corrigido))

Antes:
[[ 3000  1000  6000]
 [ 2500  2500  5500]
 [ 1000  3000  7000]
 [10000  5000 16000]]
Depois:
[[ 3000  1000  6000]
 [ 2500  2500  5500]
 [ 1000  3000 10000]
 [10000  5000 16000]]


In [9]:
"""
E) Recalculando o endividamento total
"""
endividamento_total = ((nd_financeiros_corrigido[:,0] + nd_financeiros_corrigido[:,1]) / nd_financeiros_corrigido[:,2] * 100)
endividamento_total 

array([66.66666667, 90.90909091, 40.        , 93.75      ])

In [10]:
"""
Bonus) Adicionando a coluna porcentual de endividamento a tabela 
"""
nd_financeiros_final = np.column_stack((nd_financeiros_corrigido, endividamento_total))
nd_financeiros_final

array([[ 3000.        ,  1000.        ,  6000.        ,    66.66666667],
       [ 2500.        ,  2500.        ,  5500.        ,    90.90909091],
       [ 1000.        ,  3000.        , 10000.        ,    40.        ],
       [10000.        ,  5000.        , 16000.        ,    93.75      ]])

Tabela Atualizada:

| custo fixo | dívida financeira | renda familiar | porcentual de endividamento
|:----:|:----:|:---|:---|
| 3000  | 1000 | 6000 |66.67
| 2500  | 2500 | 5500 |90.91
| 1000  | 3000 | 10000 |40
| 10000 | 5000 | 16000 |93.75

# 4.  É muito comum precisarmos identificar valores especiais e darmos tratamento a eles quer seja alterando-os quer seja descartando-os. <a name="4"></a>

<div style="text-align: right"
     
[Voltar ao índice](#questoes)

O trecho de código abaixo gera um ndarray com números pseudo aleatórios. Considere que para efeitos do estudo que virá, devemos desconsiderar valores iguais a zero. Sendo assim:

i) crie um objeto ```bool_zero``` que traga uma sequencia de booleanos do mesmo tamanho que o objeto poi, e que vale ```True``` quando o valor de ```poi``` é zero, e ```False``` caso contrário.

ii) Conte quantos valores zero existem. Lembre-se de que no final das contas, ```True``` vale 1 para o Python, e ```False``` vale zero, então uma boa dica seria usar a função ```sum()```.

iii) Utilize a indexação booleana que você aprendeu para criar uma variável ```poi_nao_zero``` que aponta para os elementos de ```poi``` diferentes de zero. Dica: você vai pode inverter os elementos do objeto que criou em *ii)* ou escrever a comparação adequada.

In [11]:
"""
Objeto poi - obs: o comando np.random.seed(1234) garante que o mesmo resultado será gerado sempre
"""
np.random.seed(1234)
poi = np.random.poisson(3, 100)
poi

array([3, 3, 6, 2, 4, 5, 4, 4, 2, 0, 3, 3, 2, 8, 2, 3, 3, 3, 0, 4, 4, 2,
       6, 6, 0, 3, 4, 3, 5, 2, 1, 3, 2, 2, 3, 2, 4, 4, 5, 7, 3, 5, 3, 3,
       2, 2, 3, 6, 2, 2, 4, 3, 5, 2, 5, 3, 0, 2, 1, 1, 4, 4, 4, 6, 2, 4,
       1, 2, 5, 5, 3, 4, 1, 4, 1, 1, 7, 1, 2, 1, 3, 7, 0, 4, 4, 0, 6, 1,
       3, 4, 2, 5, 4, 5, 5, 3, 4, 6, 8, 2])

In [12]:
"""
A) Criando o objeto bool_zero 
"""
bool_zero = poi == 0 
bool_zero

array([False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,  True, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False])

In [13]:
"""
B) Somando a quantidade de zeros dos elementos de bool_zero
"""
bool_zero.sum()

6

In [14]:
"""
C) Utilizando a indexação booleana para criar uma seleção de não-zeros
"""
bool_nao_zero = poi >= 1 
bool_nao_zero

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [15]:
"""
Bonus) Somando a quantidade de zeros dos elementos de bool_nao_zero
"""
bool_nao_zero.sum()

94

In [16]:
"""
Conferindo valores:

 A função Len retorna quantos elementos tem o objeto poi
"""
print('A quantidade de elementos no objeto "poi" é {};'.format(len(poi)))

"""
Soma dos elementos no objeto bool_nao_zero
"""
print('A quantidade de elementos no objeto "bool_nao_zero" é {};'.format(bool_nao_zero.sum()))

"""
Soma dos elementos no objeto bool_zero
"""
print('A quantidade de elementos no objeto "bool_zero" é {}.'.format(bool_zero.sum()))

A quantidade de elementos no objeto "poi" é 100;
A quantidade de elementos no objeto "bool_nao_zero" é 94;
A quantidade de elementos no objeto "bool_zero" é 6.
